src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseSpanBasedStringConcat.cs

```diff
@@ -0,0 +1,310 @@
+            }
+        }
+
+        internal static ImmutableArray<IOperation> FlattenBinaryOperation(IBinaryOperation root)
+        {
+            var stack = new Stack<IBinaryOperation>();
+            var builder = ImmutableArray.CreateBuilder<IOperation>();
+            GoLeft(root);
+
+            while (stack.Count != 0)
+            {
+                var current = stack.Pop();
+
+                if (current.LeftOperand is not IBinaryOperation leftBinary || leftBinary.OperatorKind != root.OperatorKind)
+                {
+                    builder.Add(current.LeftOperand);
+                }
+
+                if (current.RightOperand is not IBinaryOperation rightBinary || rightBinary.OperatorKind != root.OperatorKind)
+                {
+                    builder.Add(current.RightOperand);
+                }
+                else
+                {
+                    GoLeft(rightBinary);
+                }
+            }
+
+            return builder.ToImmutable();
+
+            void GoLeft(IBinaryOperation operation)
+            {
+                IBinaryOperation? current = operation;
+                while (current is not null && current.OperatorKind == root.OperatorKind)
+                {
+                    stack.Push(current);
+                    current = current.LeftOperand as IBinaryOperation;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Remove the built in implicit conversion on operands to concat.
+        /// In VB, the conversion can be to either string or object.
+        /// In C#, the conversion is always to object.
+        /// </summary>
+        internal static IOperation WalkDownBuiltInImplicitConversionOnConcatOperand(IOperation operand)
+        {
+            if (operand is not IConversionOperation conversion)
+                return operand;
+            if (!conversion.IsImplicit || conversion.Conversion.IsUserDefined)
+                return conversion;
+
+            switch (conversion.Language)
+            {
+                case LanguageNames.CSharp when conversion.Type.SpecialType is SpecialType.System_Object:
+                case LanguageNames.VisualBasic when conversion.Type.SpecialType is SpecialType.System_Object or SpecialType.System_String:
+                    return conversion.Operand;
+                default:
+                    return conversion;
+            }
+        }
+
+        // Use readonly struct instead of record type to save on allocations, since it's not passed by-value.
+        // We aren't comparing these.
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+        internal readonly struct RequiredSymbols
+#pragma warning restore CA1815 // Override equals and operator equals on value types
+        {
+            private RequiredSymbols(
+                INamedTypeSymbol stringType, INamedTypeSymbol roscharType,
+                IMethodSymbol substring1, IMethodSymbol substring2,
```

Comment:
Kindly use descriptive symbol names
```suggestion
                IMethodSymbol substringStart, IMethodSymbol substringStartLength,
```

Reply:
Fixed